### DISCUSSION TITLE
Understanding the difference of using await screen.findBy and await waitFor

### DISCUSSION THREAD

**Discussion by orchidenver** [mswjs/msw#1614] [Q&A]
In order to learn MSW I created a simple APP:
```typescript
interface User {
  userId: number;
  id: number;
  title: string;
  body: string;
}

const getData = async (setState: Dispatch<SetStateAction<User | null>>) => {
  try {
    const resp = await fetch("https://jsonplaceholder.typicode.com/posts/1");
    const data: User = await resp.json();
    setState(data);
  } catch (error) {}
};

function App() {
  const [first, setfirst] = useState<string>("");
  const [user, setUser] = useState<User | null>({
    userId: 45,
    id: 56,
    title: "hello",
    body: "ooo",
  });

  useEffect(() => {
    getData(setUser);
  }, []);

  return (
    <>
      {user?.id && <h1>{user?.title}</h1>}
      <form>
        <label htmlFor="test">Test</label>
        <input
          type="text"
          id="test"
          value={first}
          onChange={(e) => setfirst(e.target.value)}
          placeholder="test"
          name="test"
        />
      </form>
    </>
  );
}
``` 
Then I created handlers, server and setup tests:
```typescript
export const handlers = [
  rest.get("https://jsonplaceholder.typicode.com/posts/1", (req, res, ctx) => {
    return res(
      ctx.status(200),
      ctx.json({
        userId: 1,
        id: 1,
        title: "bye",
        body: "quia et suscipit suscipit recusandae consequuntur expedita et cum reprehenderit molestiae ut ut quas totam nostrum rerum est autem sunt rem eveniet architecto",
      })
    );
  }),
];

export const server = setupServer(...handlers);

import "@testing-library/jest-dom";
import { server } from "./mocks/server";
beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());
``` 

My test fails when I try to check the value of headText variable, but also works with waitFor:
```typescript
test("API call", async () => {
  render(<App />);
  const heading = await screen.findByRole("heading");
  const headText = heading.textContent;
  expect(heading).toBeInTheDocument();
  expect(headText).toBe("bye"); // expected bye, receiver hello
  await waitFor(() => {
    const titie = screen.getByRole("heading").textContent;
    expect(titie).toBe("bye");  // expected bye, receiver bye
  });
});
``` 

If I change in my App initial state of user to null, then both await screen.findBy and await waitFor work correctly.
This behavior confused me so I try to find an explanation of such behavior.

**Comment by kettanaito** [comment]
Hi, @orchidenver. I think the behavior you're experiencing is correct. It's not specific to MSW in any way. 

What happens is this:

1. Your `App` renders the first time, using the initial value of the state (the heading is "hello").
2. Then, React runs all `useEffect`s. This triggers the `getData()` asynchronous side effect that takes time to resolve.
3. The `getData()` side effect resolves and updates the state as a result. This causes React to re-render your component and you get "bye" in the UI. 

That's why you get "hello" initially in your test (the first render, and any subsequent renders until `getData()` resolves). That's absolutely correct and you need to account for that in your tests. 

To test asynchronous changes to the UI, either use the `findBy*` API or `waitFor`—both are equally valid approaches. I'd recommend `findBy*` as the primary choice because it forces you to write semantic markup.

**Reply by orchidenver** [reply]
Hello, thenks for the reply.

The weird thing for me was that in my case **findBy** finds the value that component has when App renders for the first time - hello. But when I use **waitFor** I receive what I expect - bye.

Also if I change my initial state **user** to **null** - both **findBy** and **waitFor** work the same way and received value equals expected one.

**Reply by kettanaito** [reply]
That's also expected. `findBy*` will grab _the first match_, which in your case is the first state. So assert the value is "hello" and then write `finyBy*` for "goodbye" and that should work.

**Reply by orchidenver** [reply]
So I changed my initial state of user:
```typescript
const [user, setUser] = useState<User | null>({
    userId: 11,
    id: 11,
    title: "bye-bye",
    body: "quia et suscipit suscipit recusandae consequuntur expedita et cum reprehenderit molestiae ut ut quas totam nostrum rerum est autem sunt rem eveniet architecto",
  });
``` 
Then I rewrite the test as you suggested and get this:
```typescript
test("API call", async () => {
  render(<App />);
  const initialHead = screen.getByRole("heading");
  expect(initialHead).toBeInTheDocument();
  expect(initialHead.textContent).toBe("bye-bye");

  const heading = await screen.findByRole("heading");
  const headText = heading.textContent;
  expect(heading).toBeInTheDocument();
  expect(headText).toBe("bye");

// API call
// -----
// Error: expect(received).toBe(expected) // Object.is equality
// Expected: "bye"
// Received: "bye-bye"
});
```
